Issue #123 view full message

HendrikPreuter opened the pull request:
Issue #123 

Done:
-expandable on click
-Tests

The message is still decoupled from the leader icon to allow for a modal to come in to show more details about the leader (ie. message history, bigger profile pic, maybe short bio) Issue #64

HendrikPreuter committed:
Show full message when clicked on

Drvanon commented:
This should be a modal, not an alert.

HendrikPreuter committed:
Remove size=auto as it does nothing

removing unused bit of code

HendrikPreuter committed:
Added expandable component in home

HendrikPreuter committed:
Added expandable component in home

Drvanon commented on gridt/src/app/home/home.page.html:
```diff
@@ -72,7 +72,7 @@ <h3>
                           <app-expandable [expandedHeight]="itemExpandedHeight" [expanded]="itemExpanded">
                             <ion-text *ngIf="itemExpanded">
-                              <p class="overflow-visible" size="auto">{{ leader.last_signal.message }}</p>
+                              <p class="overflow-visible">{{ leader.last_signal.message }}</p>
```
I personally really liked the idea of having to tap on the message to see it. That would mean making this a conditional statement.

HendrikPreuter replied:
That should not be there as it does nothing.
And i agree with you, that is what it does already!

HendrikPreuter committed:
Added text expanding

Drvanon commented:
Do you want feedback already?

HendrikPreuter commented on gridt/src/app/home/home.page.html:
```diff
@@ -59,9 +59,23 @@ <h3>
+                        </ion-button> -->
+                        <ion-text *ngIf="!itemExpanded">
+                          <p class="overflow-elipsis">{{ leader.last_signal.message }}</p>
+                        </ion-text>
```
This piece shows the cut-off version of the text. if the text is clicked showFullMessage is executed which toggles itemExpanded which then in turn disables this code and enables the part you comented on below.

Drvanon committed:
Update CI.yml (#139)

* Update CI.yml

* Update CI.yml

We would like to make absolutely sure that master is testing positive and pushes are rare, so this is not costly. Pull requests to development are what is most important, but when they succeed that will be the new state of the development branch, so we won't need to test that as well.

marickmanrho committed:
Create FUNDING.yml (#140)

HendrikPreuter reviewed (commented):
> Do you want feedback already?

If you want to, feel free. Tho i am still working on the tests

HendrikPreuter committed:
adding tests

HendrikPreuter committed:
Merge branch 'Issue-123-ViewFullMessage' of https://github.com/GridtNetwork/gridt-client into Issue-123-ViewFullMessage

HendrikPreuter committed:
Merge branch 'development' of https://github.com/GridtNetwork/gridt-client into Issue-123-ViewFullMessage

HendrikPreuter committed:
Git merge with development

HendrikPreuter committed:
Merge branch 'development' of https://github.com/GridtNetwork/gridt-client into Issue-123-ViewFullMessage

HendrikPreuter committed:
Restructure of expandable component

Drvanon commented on gridt/src/app/home/expandable/expandable.component.html:
```diff
@@ -0,0 +1,5 @@
+<div class='expand-wrapper' [class.expanded]="itemExpanded" (click)="toggleFullMessage()">
+  <ion-text>
+    <p class="overflow-visible">{{signalMessage}}</p>
```
Style comment: `{{ signalMessage }}` is nicer than `{{signalMessage}}`.

Drvanon commented on gridt/src/app/home/expandable/expandable.component.scss:
```diff
@@ -0,0 +1,58 @@
+        }
+}
+
+@media screen and (min-width: 281px) and (max-width: 370px) {
+        .expanded {
+                min-height: 150px;
+
+        }
+}
+@media screen  and (min-width: 371px) and (max-width: 400px) {
+        .expanded {
+                min-height: 120px;
+        }
+} 
+@media screen  and (min-width: 401px) and (max-width: 500px) {
+        .expanded {
+                min-height: 110px;
+        }
+} 
+@media screen  and (min-width: 501px) and (max-width: 600px) {
+        .expanded {
+                min-height: 70px;
+        }
+} 
+@media screen  and (min-width: 601px) and (max-width: 800px) {
+        .expanded {
+                min-height: 100px;
+        }
+} 
+
+@media screen  and (min-width: 801px) and (max-width: 1500px) {
+        .expanded {
+                min-height: 120px;
+        }
+} 
+@media screen  and (min-width: 1501px) {
+        .expanded {
+                min-height: 70px;
+        }
+} 
```
Generally speaking it is nicer to solve this with the classes from ionic, though if there is no other option this works, it's just hard to do maintenance on.

Drvanon replied:
Hmm, it might not be possible to do it differently.

HendrikPreuter replied:
I agree, however i haven't found a way just yet. But i can fix the other issues and leave this over the weekend and just keep this pull request unresolved.

Drvanon replied:
Partial solution: 
```scss
.expand-wrapper p {
        transition:max-height 0.5s;
        max-height: 16px;
        height: auto;
        overflow: hidden;
        text-overflow: ellipsis;
}
.expanded p {
        max-height: 300px;
        white-space: normal;
}
```

Drvanon commented on gridt/src/app/home/expandable/expandable.component.ts:
```diff
@@ -0,0 +1,24 @@
+  templateUrl: "./expandable.component.html",
+  styleUrls: ["./expandable.component.scss"],
+})
+export class ExpandableComponent implements OnInit {
```
I think a more appropriate name for this could be chosen. The `Component` in a component class is generally used to indicate that it _is_ a component, not to describe what type of component it is. A more name that would fit convention better could be something like `SignalMessageComponent`.

Drvanon commented on gridt/src/app/home/expandable/expandable.component.ts:
```diff
@@ -0,0 +1,24 @@
+      this.itemExpanded = !this.itemExpanded;
+    }
+  }
+
+
```
Redundant white space.

Drvanon commented on gridt/src/app/home/expandable/expandable.component.ts:
```diff
@@ -0,0 +1,24 @@
+
+  ngOnInit() {}
+
+  async toggleFullMessage() {
```
Why is this async?

HendrikPreuter replied:
Shouldn't be, leftover from an earlier version. Good catch.

Drvanon commented on gridt/src/app/home/expandable/expandable.component.ts:
```diff
@@ -0,0 +1,24 @@
+  @Input() signalMessage;
+  itemExpanded = false;
+
+  constructor() { }
+
+  ngOnInit() {}
```
You can leave these two functions as well as the implementation of ngOnInit entirely.

Drvanon commented on gridt/src/app/home/expandable/expandable.component.spec.ts:
```diff
@@ -0,0 +1,54 @@
+    expect(component).toBeTruthy();
+  });
+
+  it('should toggle itemExpanded', () => {
```
Good test.

Drvanon replied:
If you are testing this, it would be sensible to test that it also does _not_ do this when the text is too short.

HendrikPreuter replied:
When i change this test:

```
  it('should toggleFullMessage on click', (done) => {
    spyOn(component, 'toggleFullMessage');
    component.signalMessage = "this message is long enough to require more space.";
    component.width = 5;

    let button = fixture.debugElement.query(By.css('.expand-wrapper')).nativeElement;
    button.click();

    fixture.whenStable().then(() => {
      expect(component.toggleFullMessage).toHaveBeenCalled();
      expect(component.itemExpanded).toBe(true, 'true after toggle');
    });
    done();
  });
```

it fails on this:

`expect(component.itemExpanded).toBe(true, 'true after toggle');`

But i don't know why

HendrikPreuter replied:
The new toggle function:
```
  ngOnInit() {
    this.width = document.getElementById('signalText').offsetWidth;
  }

  toggleFullMessage() {
    if (this.signalMessage.length > this.width) {
      this.itemExpanded = !this.itemExpanded;
    }
  }
```

Drvanon replied:
https://hatoum.com/blog/2016/11/12/jasmine-unit-testing-dont-forget-to-callthrough

Drvanon commented on gridt/src/app/home/expandable/expandable.component.spec.ts:
```diff
@@ -0,0 +1,54 @@
+    expect(component.itemExpanded).toBe(false, 'false after second toggle');
+  });
+
+  it('should toggleFullMessage on click', (done) => {
+    spyOn(component, 'toggleFullMessage');
+
+    let button = fixture.debugElement.query(By.css('.expand-wrapper')).nativeElement;
+    button.click();
+
+    fixture.whenStable().then(() => {
+      expect(component.toggleFullMessage).toHaveBeenCalled();
+    });
+    done();
+  });
```
Excellent! You are starting to understand the tests!

HendrikPreuter committed:
Added tests

Drvanon commented on gridt/src/app/home/expandable/expandable.component.ts:
```diff
@@ -0,0 +1,24 @@
+  ngOnInit() {}
+
+  async toggleFullMessage() {
+    if(this.signalMessage.length > 10) {
```
This is rather arbitrary choice. Are you able to find out if the message is being wrapped or not and then use that information to perform the check?

Drvanon commented on gridt/src/app/home/home.page.html:
```diff
@@ -59,9 +59,24 @@ <h3>
-                      </small>
+                      <div *ngIf="isLeaderDone(leader, movement) && leader.last_signal.message" id="expandable">
+                        <app-expandable signalMessage={{leader.last_signal.message}}></app-expandable>
+                        <!-- <ion-button fill="clear" (click)="showFullMessage(leader.last_signal.message)">
+                          <div class="leader-message">
+                            <small>
+                              {{ leader.last_signal.message }}
+                            </small>
+                          </div>
+                        </ion-button> -->
+                        <!-- <ion-text *ngIf="!itemExpanded">
+                          <p class="overflow-elipsis">{{ leader.last_signal.message }}</p>
+                        </ion-text>
+                          <app-expandable [signalMessage]="leader.last_signal.message">
+                            <ion-text *ngIf="itemExpanded">
+                              <p class="overflow-visible">{{ leader.last_signal.message }}</p>
+                            </ion-text>
+                          </app-expandable> -->
```
General rule of thumb: never commit commented code. It serves no purpose as your history is already captured by git.

Drvanon commented on gridt/src/app/home/home.page.scss:
```diff
@@ -1,3 +1,7 @@
+.signal {
+
+}
```
This does not achieve anything.

Drvanon commented on gridt/src/app/home/expandable/expandable.component.scss:
```diff
@@ -0,0 +1,58 @@
+        min-height: 150px;
+}
+
+@media screen and (max-width: 280px) {
+        .expanded {
+                min-height: 310px;
+        }
+}
+
+@media screen and (min-width: 281px) and (max-width: 370px) {
+        .expanded {
+                min-height: 150px;
+
+        }
+}
+@media screen  and (min-width: 371px) and (max-width: 400px) {
+        .expanded {
+                min-height: 120px;
+        }
+} 
+@media screen  and (min-width: 401px) and (max-width: 500px) {
+        .expanded {
+                min-height: 110px;
+        }
+} 
+@media screen  and (min-width: 501px) and (max-width: 600px) {
+        .expanded {
+                min-height: 70px;
+        }
+} 
+@media screen  and (min-width: 601px) and (max-width: 800px) {
+        .expanded {
+                min-height: 100px;
+        }
+} 
+
+@media screen  and (min-width: 801px) and (max-width: 1500px) {
+        .expanded {
+                min-height: 120px;
+        }
```
Is this even necessary? When I click on the message, I want to see the message. It should not take up more space than necessary.

HendrikPreuter replied:
I agree that it shouldn't take up more room than necessary. That's why i added so many different versions. So that on each screen type the wasted space on expansion is minimized.

Drvanon replied:
Isn't the overflow visible class what makes the ... appear? Maybe you could just remove that class conditionally.

HendrikPreuter replied:
```
.overflow-ellipsis {
        text-overflow: ellipsis;
        white-space: normal;
        overflow: hidden;
    }
```
Should provide the '...' however they don't appear for me, whatever i do.
```
.overflow-visible {
        white-space: initial;
        
    }
```

Is what makes it that the text isn't cut off and continue on a new line.
If i remove the part commented on in the scss file than yes, it still expands. However the smooth transitional animation is thereby also removed as that only seems to work if the min-height changes.

So then the question becomes do we want easier to maintain or a nicer view experience for the user?

Drvanon commented on gridt/src/app/home/expandable/expandable.component.spec.ts:
```diff
@@ -0,0 +1,54 @@
+  });
+
+  it('should show the expanded view', () => {
+    component.signalMessage = "this message is long enough to require more space.";
+    component.toggleFullMessage();
+    expect(component.itemExpanded).toBe(true, 'true after toggle');
+    let expandedEl = fixture.debugElement.query(By.css('.expand-wrapper, .expanded')).nativeElement;
+    expect(expandedEl.innerHTML).not.toBeNull();
```
Good idea to test this.

Drvanon commented on gridt/src/app/home/home.page.spec.ts:
```diff
@@ -1,6 +1,6 @@
 import { of } from 'rxjs';
```
Changes here are unecessary.

Drvanon commented on gridt/src/app/home/home.page.ts:
```diff
@@ -14,6 +14,7 @@ import { SwapService } from '../core/swap.service';
 })
 export class HomePage implements OnInit, OnDestroy {
   movements$ = new Observable<Movement[]>();
+  itemExpanded = false;
```
This should be removed.

Drvanon reviewed (changes requested):
You are definitely reaching the final stages of this commit. I like the way it is designed very much now, but there is some cleanup to do.

HendrikPreuter committed:
Add rework based on feedback

HendrikPreuter committed:
Merge branch 'development' of https://github.com/GridtNetwork/gridt-client into Issue-123-ViewFullMessage

HendrikPreuter committed:
Remove unused imports

marickmanrho reviewed (approved):
I think the solution presented is very elegant.

Drvanon commented on gridt/src/app/home/home.page.html:
```diff
@@ -59,9 +59,9 @@ <h3>
-                        {{ leader.last_signal.message }}
-                      </small>
+                      <div *ngIf="isLeaderDone(leader, movement) && leader.last_signal.message" id="expandable">
+                        <app-expandable signalMessage={{leader.last_signal.message}}></app-expandable>
```
I would like there to be a different name here

HendrikPreuter replied:
Is leadersMessage a better name?

HendrikPreuter replied:
Resolved via what's app.

Drvanon commented on gridt/src/app/home/expandable/expandable.component.spec.ts:
```diff
@@ -0,0 +1,82 @@
+      expect(component.toggleFullMessage).toHaveBeenCalled();
+      expect(component.itemExpanded).toBe(true, 'true after toggle');
+    });
+    done();
```
This done should be in the promise, now it isn't doing anything.

Drvanon commented on gridt/src/app/home/expandable/expandable.component.spec.ts:
```diff
@@ -0,0 +1,82 @@
+      expect(component.toggleFullMessage).toHaveBeenCalled();
+      expect(component.itemExpanded).toBe(false, 'not toggled');
+    });
+    done();
```
Same for this done

Drvanon reviewed (commented):
Really good commit, it's starting to come together beautifully. I have to still look at the functionality when I get out of the hospital, but the code is looking very solid!

HendrikPreuter committed:
Move done inside the promise

HendrikPreuter committed:
Change name of signalMessage to leadersMessage

HendrikPreuter committed:
Change signalMessage to signalText

HendrikPreuter committed:
Change selector expandable component

Drvanon commented on gridt/src/app/home/expandable/expandable.component.ts:
```diff
@@ -0,0 +1,22 @@
+})
+export class SignalMessageComponent implements OnInit{
+
+  @Input() signalText;
```
Within the context of our application I think a more suitable name is message text. Very banal change, but I think it helps comprehension.

HendrikPreuter replied:
Why is that? Every time someone talked about giving a message to the followers it was called signaling. So why is signal message worse then message text?

Drvanon commented on gridt/src/app/home/expandable/expandable.component.scss:
```diff
@@ -0,0 +1,17 @@
+.p {
```
I need a computer to be able to check out the functionality of this. But i won't be home until Thursday. Maybe you could make me a gif so I can see it?

HendrikPreuter replied:
yes, i will send you one soon.

HendrikPreuter replied:
It is in the slack/client

Drvanon commented on gridt/src/app/home/expandable/expandable.component.spec.ts:
```diff
@@ -0,0 +1,82 @@
+
+  it('should toggle itemExpanded', () => {
+    component.signalText = "this message is long enough to require more space.";
+    component.width = 5;
```
By setting the width yourself you are not checking the width determining code you have created. Is it possible to set the screen width in the test frame work?

HendrikPreuter replied:
That is true! I hadn't thought of that. I will be keeping these tests i think and try and write 2 more to specifically test that if it can be set.

HendrikPreuter replied:
So far, i haven't found a way to do this.
However even if we get the full screen width, that doesn't get us there as we need the width of <p>

HendrikPreuter replied:
I could expect width to be bigger than 0, that way we know the function in onInnit did something. But that doesn't help much as it could still be a wrong value.

Drvanon commented on gridt/src/app/home/expandable/expandable.component.ts:
```diff
@@ -0,0 +1,22 @@
+import { Component, OnInit, Input } from "@angular/core";
```
The file and folder name should be renamed still.

HendrikPreuter replied:
this has given me some issues yesterday, but i will try again today.

HendrikPreuter replied:
Turns out, i forgot to redo ionic serve...

HendrikPreuter committed:
Update name

Drvanon commented on gridt/src/app/home/home.page.html:
```diff
@@ -59,9 +59,9 @@ <h3>
-                        {{ leader.last_signal.message }}
-                      </small>
+                      <div *ngIf="isLeaderDone(leader, movement) && leader.last_signal.message" id="expandable">
+                        <signal-message signalText={{leader.last_signal.message}}></signal-message>
```
This element has only one real input and it does not necessarily need to be named, so what you could strive for (unless you have a good counter argument) is:

`<signal-message>{{ leader.last_signal.message }}</signal-message>`

HendrikPreuter replied:
that is what i had initially, but then i needed to switch over all of  the functionality for the switching the view to the signal message component. If i were to do it like this i will have to replace the <p> in the signal message component with <ng-content> and i am not sure the styling would work with that as good as it does now.

Drvanon replied:
I feel like something went wrong when you typed this message. I also don't think I completely understand the argument, would you care to elaborate?

HendrikPreuter committed:
Update style

HendrikPreuter reviewed (commented):
I will be processing the feedback today.

HendrikPreuter committed:
Add style so that both icons stay stationary and renaming of signal-message

HendrikPreuter committed:
Add missing files

Yori-O92 committed:
Update README.md

AntoninThioux committed:
Merge branch 'development-SME', remote-tracking branch 'origin' into Issue-123-ViewFullMessage

AntoninThioux committed:
accidentally uploaded some files
